### ISSUE TITLE
getway

### ISSUE THREAD

**Issue by waarpcncjh** [waarp/Waarp#2]
hello When I was learning AllJarsWaarpR66-3.0.4, has launched the service side (server. Bat) and client (R66ClGUI. Bat), can also succeed through the client send the file to the server, but now also don't know how to put the gateway (getway) is applied to the inside, please help me, how in the process of a transfer application gateway, the gateway how to start, how to configure, how to put forward a transfer requests through a gateway to the server

**Comment by waarpcncjh** [comment]
[WaarpGatewayKernel-master.zip ]This is the gateway R66 code? , how to start his, how to configure his relationship with the client, server

**Comment by waarpcncjh** [comment]
https://github.com/waarp/WaarpR66/wiki/DMZ-Usage，{By installing a DMZ R66 server, through the rules, it could act as a forward request will full checking.}Inside, there are words, 【through the rules,】what is the rule, how to configure the rule

**Comment by fredericBregier** [comment]
The DMZ is through R66Proxy: see here http://waarp.github.io/Waarp/R66Proxy.html
The Gateway Waarp modules are sub modules or sub projects that can interact with a Waarp R66 server. While you can use only Waarp R66 server and client (from the very same Waarp R66 distribution), you can also have some clients using FTP for instance. This is where Waarp Gateway FTP is coming.

Waarp Gateway FTP is a gateway between FTP client (not R66 client) to (for instance, because you can do many other things with it) Waarp R66 server.
The config directory (in http://sourceforge.net/projects/waarp/files/AllJars/AllJarsWaarpGatewayFtp-3.0.2.zip/download ) contains some examples of configurations you can based on.
The rest is explained (and sub pages as API) there: http://waarp.github.io/Waarp/WaarpGatewayFtp.html

**Comment by waarpcncjh** [comment]
I don't know how to configure, gateway forwarding rules, please give me an example of a gateway rules!

**Comment by fredericBregier** [comment]
See:
1) https://github.com/waarp/WaarpGatewayFtp/tree/master/src/main/example/config
(available in distribution in config)
2) http://waarp.github.io/WaarpR66/api/org/waarp/gateway/kernel/exec/AbstractExecutor.html where the various execute commands available are described.
3) http://waarp.github.io/WaarpR66/api/org/waarp/gateway/kernel/exec/R66PreparedTransferExecutor.html where the specific R66PreparedTransferExecutor is explained.

Here are the copy of the texts.

Abstract Executor class. If the command starts with "REFUSED", the command will be refused for execution. If "REFUSED" is set, the command "RETR" or "STOR" like operations will be stopped at starting of command.
If the command starts with "EXECUTE", the following will be a command to be executed.
If the command starts with "JAVAEXECUTE", the following will be a command through Java class to be executed.
If the command starts with "R66PREPARETRANSFER", the following will be a r66 prepare transfer execution (asynchronous operation only).
The following replacement are done dynamically before the command is executed:
- #BASEPATH# is replaced by the full path for the root of FTP Directory
- #FILE# is replaced by the current file path relative to FTP Directory (so #BASEPATH##FILE# is the full path of the file)
- #USER# is replaced by the username
- #ACCOUNT# is replaced by the account
- #COMMAND# is replaced by the command issued for the file
- #SPECIALID# is replaced by the FTP id of the transfer (whatever in or out)
- #UUID# is replaced by a special UUID globally unique for the transfer, in general to be placed in -info part (for instance ##UUID## giving #uuid#)

R66PreparedTransferExecutor class. If the command starts with "REFUSED", the command will be refused for execution. If "REFUSED" is set, the command "RETR" or "STOR" like operations will be stopped at starting of command. Format is like r66send command in any order except "-info" which should be the last item:
"-to Host -file FILE -rule RULE [-md5] [-nolog] [-start yyyyMMddHHmmss or -delay (delay or +delay)] [-info INFO]"

INFO is the only one field that can contains blank character.

The following replacement are done dynamically before the command is executed:
- #BASEPATH# is replaced by the full path for the root of FTP Directory
- #FILE# is replaced by the current file path relative to FTP Directory (so #BASEPATH##FILE# is the full path of the file)
- #USER# is replaced by the username
- #ACCOUNT# is replaced by the account
- #COMMAND# is replaced by the command issued for the file
- #SPECIALID# is replaced by the FTP id of the transfer (whatever in or out)
- #UUID# is replaced by a special UUID globally unique for the transfer, in general to be placed in -info part (for instance ##UUID## giving #uuid#)

So for instance "-to Host -file #BASEPATH##FILE# -rule RULE [-md5] [-nolog] [-delay +delay] [-info ##UUID## #USER# #ACCOUNT# #COMMAND# INFO]"
will be a standard use of this function.


So in the main xml configuration file, you can change:

    <exec>
        <retrievecmd>NONE EXECUTE J:\GG\testexec.bat</retrievecmd>
        <storecmd>NONE EXECUTE J:\GG\testexec.bat</storecmd>
    </exec>

by what you need, for instance:

    <exec>
        <retrievecmd>R66PREPARETRANSFER -to Host -file #BASEPATH##FILE# -rule RULERECV</retrievecmd>
        <storecmd>R66PREPARETRANSFER -to Host -file #BASEPATH##FILE# -rule RULESTOR</storecmd>
    </exec>

**Comment by waarpcncjh** [comment]
First of all thank you for your patience to answer, but I still don't understand, in the web site [https://github.com/waarp/WaarpR66/wiki/DMZ-Usage], you put forward three DMZ solution, I asked the question is one of the application of the second option, but do not know why, every time you give me the answer, have the shadow of the FTP, our application is to use TCP, HTTP transport, I want to know is in the case of does not apply to your FTP gateway, use your second solution [Waarp R66 in the forward mode

**Comment by fredericBregier** [comment]
OK, so for the DMZ usage, you don't need at all Waarp Gateway FTP. I probably do not understand your question at first time.  My bad...

You need Waarp R66 Proxy, which is already in the zip package Waarp R66.
See here http://waarp.github.io/Waarp/R66Proxy.html

The Waarp R66 Proxy server is like a normal R66 server, except there is no database.

The configuration is found as an example in the distribution under directory named "proxy/config" or here on github: https://github.com/waarp/WaarpProxyR66/tree/master/src/main/config

The configuration must have the authentification part, as in the R66 server, but the rules are not necessary (since they are deduced by final R66 servers). But you need to setup the "proxy part" : see "<network><serverproxy>". You need to specify how the proxy runs there (source address:port - target address:port). Then on you remote server, you should use the defined source address:port as the new sibling for the proxified server.

So let say Client A connect to Server B through Proxy C.
In C, you say the Proxy is listening on address:port X:port from outside and will forward to address:port Y:p' which is the real address:port of Server B.
Client A will have to connect to X:port and configuring Server B address as X:port, such that the client will in fact connect to Proxy relaying to Server B transparently.

**Comment by waarpcncjh** [comment]
If you use a proxy pattern, if there are 1000 server in the network, and outside network to access the 1000 servers to access through a proxy, it brings a few problems.

The first: according to the rules of proxy configuration file, there will be 2000 in the configuration file < serverproxy > tag, also means that the proxy server to monitor 2000 local port. Would it too much?

The second: the configuration content also have to the proxy server is configured before starting?

**Comment by waarpcncjh** [comment]
What is the meaning of a passage from below, I didn't understand his usage,

 [


2) Waarp R66 in forward mode

By installing a DMZ R66 server, through the rules, it could act as a forward request will full checking.

The interest is to have full checking at once for all type of transfers, without having to directly connect to internal R66 servers from outside. The drawback is that this DMZ R66 server has a full configuration (using database and all host authentications), which could lead to some issues in very high level protected area.

    Waarp R66
        Each send file request received by the R66 DMZ server from an internal R66 partner will be followed by a send request in R66 (within R66 as a task) to a remote R66 server
        Each recv file request received by the R66 DMZ server from an internal R66 partner will be prefixed by a recv request in R66 (within R66 as a task) to a remote R66 server, then followed by the R66 internal transfer
]

**Comment by fredericBregier** [comment]
Your question is quite complex, since you bring some information but not all :-(
However, following what you said:
- the number of open ports within the proxy server is twice the number of servers to reach (whatever inside or outside): 
    - this is a "passive" proxy (no database, only target/source definition and host definitions)
    - let say there is 100 servers outside but only 10 of them would be a target of new transfer connection (from inside servers)
    - let say there is 200 servers inside but only 20 of them would be a target of new transfer connection (from outside servers)
    - then it means that 10+20 proxy connections must be setup, so 30x2 listening ports
- if you want to check the rule or forward the transfer from/to outside, then you need to install a R66 server that acts as an "active" proxy:
    - this could be a "basic" R66 server with no database (it will not keep, except through log file, any trace of transfer - in fact, it is not so true, since you can use the )
    - this could be a "standard" R66 server with a database
    - the principle will be that each rule (how a transfer is handle) will be on this "active" proxy as: get the file and forward it to someone else, getting for instance the "someone else" from the file transfer information (string passed in argument). It's up to you to define how you want to do this. Doing this way, you could have 1 rule only.
    - another principle could be to have as much as rule as much as pair of server communicating to each other (so if there is 100 outside servers and 200 inside servers, this could lead up to 100*200*2 rules (*2 due to RECV and SEND ways)). 
   - An intermediate way would be to have "common" rules between various servers (meaning you need to have the partner in the file transfer information) or "common" rules to a single server (meaning that you will have (100+200)*2 rules, one per destination and way (RECV/SEND)).

*R66 can be used in so many ways, I can't say what is best for you.*

I understand of course the "security" of the DMZ, so the reason of the "passive" proxy. But if it is not ok for you (too many open ports), then you have to find the best way for you.

For your information, the limit is on the number of port a single process can open, not on memory side, since, as long as there are not 1 millions conccurent transfers, the memory and threads are shared across all open ports, so it is already optimzed for that. The only limitations will be the limit of open ports from a single process (see your OS configuration) and of course the number of concurrent transferts since each of them brings some memory usage.

**Comment by waarpcncjh** [comment]
Once again thank you for your explanation. Our business is open port has certain limitation, so use the proxy pattern is a bit difficult. Our business is to realize the transmission of files between R66 server, the transfer rule is between any two server can transfer files. According to the R66 now the implementation of the basic can achieve our business, now, however, involves the DMZ, just a little bit of a problem, or I don't know how to use R66 to solve our problems, do not know to have a solution

**Comment by waarpcncjh** [comment]
My agent in use you provide start the broker configuration file (R66Proxy) when a mistake, I don't know how to solve, please help. The following error information 。

10:34:46.620 [main] DEBUG o.w.c.c.WaarpConstraintLimitHandler - org.waarp.common.cpu.WaarpConstraintLimitHandler.setHandler(WaarpConstraintLimitHandler.java:424) : Activate Throttle bandwidth according to CPU usage
10:34:46.629 [main] ERROR org.waarp.openr66.proxy.R66Proxy - org.waarp.openr66.proxy.R66Proxy.initialize(R66Proxy.java:67) : Startup of Proxy is in error
io.netty.channel.ChannelException: address already in use by: [id: 0x412f4b7d, local:0]
	at io.netty.channel.local.LocalChannelRegistry.register(LocalChannelRegistry.java:46) ~[netty-all-4.1.0.Beta5.jar:4.1.0.Beta5]
	at io.netty.channel.local.LocalServerChannel.doBind(LocalServerChannel.java:91) ~[netty-all-4.1.0.Beta5.jar:4.1.0.Beta5]
	at io.netty.channel.AbstractChannel$AbstractUnsafe.bind(AbstractChannel.java:522) ~[netty-all-4.1.0.Beta5.jar:4.1.0.Beta5]
	at io.netty.channel.DefaultChannelPipeline$HeadContext.bind(DefaultChannelPipeline.java:1196) ~[netty-all-4.1.0.Beta5.jar:4.1.0.Beta5]
	at io.netty.channel.ChannelHandlerInvokerUtil.invokeBindNow(ChannelHandlerInvokerUtil.java:108) ~[netty-all-4.1.0.Beta5.jar:4.1.0.Beta5]
	at io.netty.channel.DefaultChannelHandlerInvoker.invokeBind(DefaultChannelHandlerInvoker.java:214) ~[netty-all-4.1.0.Beta5.jar:4.1.0.Beta5]
	at io.netty.channel.AbstractChannelHandlerContext.bind(AbstractChannelHandlerContext.java:208) ~[netty-all-4.1.0.Beta5.jar:4.1.0.Beta5]
	at io.netty.channel.DefaultChannelPipeline.bind(DefaultChannelPipeline.java:1003) ~[netty-all-4.1.0.Beta5.jar:4.1.0.Beta5]
	at io.netty.channel.AbstractChannel.bind(AbstractChannel.java:216) ~[netty-all-4.1.0.Beta5.jar:4.1.0.Beta5]
	at io.netty.bootstrap.AbstractBootstrap$2.run(AbstractBootstrap.java:357) ~[netty-all-4.1.0.Beta5.jar:4.1.0.Beta5]
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:322) ~[netty-all-4.1.0.Beta5.jar:4.1.0.Beta5]
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:356) ~[netty-all-4.1.0.Beta5.jar:4.1.0.Beta5]
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:703) ~[netty-all-4.1.0.Beta5.jar:4.1.0.Beta5]
	at java.lang.Thread.run(Unknown Source) ~[na:1.6.0_20]

**Comment by fredericBregier** [comment]
10:34:46.629 [main] ERROR org.waarp.openr66.proxy.R66Proxy - org.waarp.openr66.proxy.R66Proxy.initialize(R66Proxy.java:67) : Startup of Proxy is in error
io.netty.channel.ChannelException: address already in use by: [id: 0x412f4b7d, local:0]

Means you try to open twice the same couple (address, port), meaning you specify twice the same couple in your configuration file of R66Proxy.

**Comment by waarpcncjh** [comment]
I don't understand, I take it out in the configuration file from the project, I also saw no repetitive, below is my configuration file content, please help me to check. 

proxy_config.xml

<?xml version="1.0" encoding="UTF-8"?>
<config xmlns:x0="http://www.w3.org/2001/XMLSchema">
    <comment>Example of config file: change its as your need.</comment>
    <identity>
        <hostid>hostproxy</hostid>
        <sslhostid>hostproxys</sslhostid>
        <cryptokey>D:\work_waarp\WaarpProxyR66-master\src\main\config\certs\test-key.des</cryptokey>
        <authentfile>D:\work_waarp\WaarpR66-master\src\main\config\OpenR66-authent-A.xml</authentfile>
    </identity>
    <server>
        <serveradmin>monadmin</serveradmin>
        <noserverpasswd>c5f4876737cf351a</noserverpasswd>
        <serverpasswdfile>D:\work_waarp\WaarpProxyR66-master\src\main\config\certs\test-passwd-http.ggp</serverpasswdfile>
        <usenossl>True</usenossl>
        <usessl>false</usessl>
        <usehttpcomp>False</usehttpcomp>
        <uselocalexec>False</uselocalexec>
        <httpadmin>D:\work_waarp\WaarpProxyR66-master\src\main\admin2</httpadmin>
        <admkeypath>D:\work_waarp\WaarpProxyR66-master\src\main\config\certs\testsslnocert.jks</admkeypath>
        <admkeystorepass>testsslnocert</admkeystorepass>
        <admkeypass>testalias</admkeypass>
        <pastlimit>86400000</pastlimit>
        <minimaldelay>5000</minimaldelay>
        <snmpconfig>D:\work_waarp\WaarpProxyR66-master\src\main\config\snmpconfig-proxy.xml</snmpconfig>
    </server>
    <network>
		<serverproxy>
			<serverlistenaddr>127.0.0.1</serverlistenaddr>
			<serverlistenport>9986</serverlistenport>
			<serverlistenssl>false</serverlistenssl>
			<serverremoteaddr>127.0.0.1</serverremoteaddr>
			<serverremoteport>6686</serverremoteport>
			<serverremotessl>false</serverremotessl>
		</serverproxy>
		<serverproxy>
			<serverlistenaddr>127.0.0.1</serverlistenaddr>
			<serverlistenport>9988</serverlistenport>
			<serverlistenssl>false</serverlistenssl>
			<serverremoteaddr>127.0.0.1</serverremoteaddr>
			<serverremoteport>6687</serverremoteport>
			<serverremotessl>true</serverremotessl>
		</serverproxy>
        <serverhttpport>10086</serverhttpport>
        <serverhttpsport>10087</serverhttpsport>
    </network>
    <ssl>
        <keypath>D:\work_waarp\WaarpProxyR66-master\src\main\config\certs\testsslnocert.jks</keypath>
        <keystorepass>testsslnocert</keystorepass>
        <keypass>testalias</keypass>
        <trustkeypath>D:\work_waarp\WaarpProxyR66-master\src\main\config\certs\testcert.jks</trustkeypath>
        <trustkeystorepass>testcert</trustkeystorepass>
        <trustuseclientauthenticate>True</trustuseclientauthenticate>
    </ssl>
    <directory>
        <serverhome>D:\work_waarp\WaarpProxyR66-master</serverhome>
        <arch>arch</arch>
        <conf>conf</conf>
    </directory>
    <limit>
        <timeoutcon>10000</timeoutcon>
        <serverthread>2</serverthread>
        <clientthread>4</clientthread>
        <memorylimit>4000000000</memorylimit>
        <sessionlimit>0</sessionlimit>
        <globallimit>0</globallimit>
        <delaylimit>10000</delaylimit>
        <usecpulimit>False</usecpulimit>
        <usejdkcpulimit>False</usejdkcpulimit>
        <cpulimit>0.9</cpulimit>
        <connlimit>0</connlimit>
        <lowcpulimit>0.5</lowcpulimit>
        <highcpulimit>0.9</highcpulimit>
        <percentdecrease>0.25</percentdecrease>
        <delaythrottle>1000</delaythrottle>
        <limitlowbandwidth>4096</limitlowbandwidth>
		<checkversion>True</checkversion>
    </limit>
</config>
snmp_config.xml

<snmpconfig>
<config>
	<localaddress>udp:0.0.0.0/2011</localaddress>
	<localaddress>tcp:0.0.0.0/2012</localaddress>
	<nbthread>4</nbthread>
	<filtered>False</filtered>
	<usetrap>True</usetrap>
	<trapinformlevel>5</trapinformlevel>
</config>
<targets>
	<target>
		<name>notificationV2c</name>
		<domain>UdpIpv4</domain>
		<address>127.0.0.1/162</address>
		<timeout>200</timeout>
		<retries>1</retries>
		<isv2>True</isv2>
	</target>
	<target>
		<name>notificationV3</name>
		<domain>UdpIpv4</domain>
		<address>127.0.0.1/162</address>
		<timeout>200</timeout>
		<retries>1</retries>
		<isv2>False</isv2>
	</target>
</targets>
<securities>
	<security>
		<securityname>SHADES</securityname>
		<securityauthprotocol>SHA</securityauthprotocol>
		<securityauthpass>SHADESAuthPassword</securityauthpass>
		<securityprivprotocol>PDES</securityprivprotocol>
		<securityprivpass>SHADESPrivPassword</securityprivpass>
	</security>
	<security>
		<securityname>TEST</securityname>
		<securityauthprotocol>SHA</securityauthprotocol>
		<securityauthpass>maplesyrup</securityauthpass>
		<securityprivprotocol>PDES</securityprivprotocol>
		<securityprivpass>maplesyrup</securityprivpass>
	</security>
	<security>
		<securityname>SHA</securityname>
		<securityauthprotocol>SHA</securityauthprotocol>
		<securityauthpass>SHAAuthPassword</securityauthpass>
	</security>
	<security>
		<securityname>MD5DES</securityname>
		<securityauthprotocol>MD5</securityauthprotocol>
		<securityauthpass>MD5DESAuthPassword</securityauthpass>
		<securityprivprotocol>PDES</securityprivprotocol>
		<securityprivpass>MD5DESPrivPassword</securityprivpass>
	</security>
	<security>
		<securityname>SHAAES128</securityname>
		<securityauthprotocol>SHA</securityauthprotocol>
		<securityauthpass>SHAAES128AuthPassword</securityauthpass>
		<securityprivprotocol>PAES128</securityprivprotocol>
		<securityprivpass>SHAAES128PrivPassword</securityprivpass>
	</security>
	<security>
		<securityname>SHAAES192</securityname>
		<securityauthprotocol>SHA</securityauthprotocol>
		<securityauthpass>SHAAES192AuthPassword</securityauthpass>
		<securityprivprotocol>PAES192</securityprivprotocol>
		<securityprivpass>SHAAES192PrivPassword</securityprivpass>
	</security>
	<security>
		<securityname>SHAAES256</securityname>
		<securityauthprotocol>SHA</securityauthprotocol>
		<securityauthpass>SHAAES256AuthPassword</securityauthpass>
		<securityprivprotocol>PAES256</securityprivprotocol>
		<securityprivpass>SHAAES256PrivPassword</securityprivpass>
	</security>
	<security>
		<securityname>MD5AES128</securityname>
		<securityauthprotocol>MD5</securityauthprotocol>
		<securityauthpass>MD5AES128AuthPassword</securityauthpass>
		<securityprivprotocol>PAES128</securityprivprotocol>
		<securityprivpass>MD5AES128PrivPassword</securityprivpass>
	</security>
	<security>
		<securityname>MD5AES192</securityname>
		<securityauthprotocol>MD5</securityauthprotocol>
		<securityauthpass>MD5AES192AuthPassword</securityauthpass>
		<securityprivprotocol>PAES192</securityprivprotocol>
		<securityprivpass>MD5AES192PrivPassword</securityprivpass>
	</security>
	<security>
		<securityname>MD5AES256</securityname>
		<securityauthprotocol>MD5</securityauthprotocol>
		<securityauthpass>MD5AES256AuthPassword</securityauthpass>
		<securityprivprotocol>PAES256</securityprivprotocol>
		<securityprivpass>MD5AES256PrivPassword</securityprivpass>
	</security>
</securities>
</snmpconfig>

**Comment by waarpcncjh** [comment]
At the time of start an agent（R66Proxy） of the main method, call serverStartup method, several main methods, as follows:

PipelineInit ();

ServerPipelineInit ();

R66Startup ();

StartHttpSupport ();

StartMonitoring ();

LaunchStatistics ();

StartRestSupport ();

One of the first pipelineInit () method to initialize a localTransaction object, he will bind an address,

ServerChannel = serverBootstrap. Bind (socketLocalServerAddress). The sync (). The channel (); .

When performing r66Startup (); Method will perform serverChannel = serverBootstrap. Bind (socketLocalServerAddress). The sync (). The channel (); . The code. Why in a service launched two monitoring service, but the true address is the same

**Comment by waarpcncjh** [comment]
As for you said I configure a number of the same address, but I do not find repeated listening address. I through the tracking code, I found mainly in the startup R66Proxy the main method of a class, there are [pipelineInit (), r66Startup () 】 two methods, they are all created LocalTransaction object, and the binding address, will have the same address (LocalAddress socketLocalServerAddress = new LocalAddress (" 0 ");) , personally, I think is that they lead to, don't know whether my guess is correct

**Comment by fredericBregier** [comment]
The Configuration is not strictly the same than in R66 Server:
- pipelineInit();
- serverPipelineInit();
- r66Startup();
- startHttpSupport();
- startMonitoring();

It might be the HTTP port of the configuration? But according to your trace, you did not reach this part since the previous log is about WaarpConstraintLimitHandler.

So I have no idea which port is already in use in your OS while you try to start R66 Proxy. You should check for instance if there is another R66 server running on the same server or another program using the very same port you specified (either for R66 protocol, either for HTTP port).

**Comment by fredericBregier** [comment]
Since you open a new issue for the very same point, closing this one.
